tests/src/exp.rs

```diff
@@ -0,0 +1,564 @@
+        Expression::and(vec![
+            Expression::eq(
+                Expression::int_bin("bin".to_string()),
+                Expression::int_val(1),
+            ),
+            Expression::eq(
+                Expression::string_bin("bin2".to_string()),
+                Expression::string_val("1".to_string()),
+            ),
+        ]),
+        &set_name,
+    );
+    let count = count_results(rs);
+    assert_eq!(count, 1, "AND Test Failed");
+
+    // OR
+    let rs = test_filter(
+        Expression::or(vec![
+            Expression::eq(
+                Expression::int_bin("bin".to_string()),
+                Expression::int_val(1),
+            ),
+            Expression::eq(
+                Expression::int_bin("bin".to_string()),
+                Expression::int_val(3),
+            ),
+        ]),
+        &set_name,
+    );
+    let count = count_results(rs);
+    assert_eq!(count, 2, "OR Test Failed");
+
+    // NOT
+    let rs = test_filter(
+        Expression::not(Expression::eq(
+            Expression::int_bin("bin".to_string()),
+            Expression::int_val(1),
+        )),
+        &set_name,
+    );
+    let count = count_results(rs);
+    assert_eq!(count, 99, "NOT Test Failed");
+}
+
+#[test]
+fn expression_data_types() {
+    let _ = env_logger::try_init();
+
+    let set_name = create_test_set(EXPECTED);
+
+    // INT
+    let rs = test_filter(
+        Expression::eq(
+            Expression::int_bin("bin".to_string()),
+            Expression::int_val(1),
+        ),
+        &set_name,
+    );
+    let count = count_results(rs);
+    assert_eq!(count, 1, "INT Test Failed");
+
+    // STRING
+    let rs = test_filter(
+        Expression::eq(
+            Expression::string_bin("bin2".to_string()),
+            Expression::string_val("1".to_string()),
+        ),
+        &set_name,
+    );
+    let count = count_results(rs);
+    assert_eq!(count, 1, "STRING Test Failed");
+
+    let rs = test_filter(
+        Expression::eq(
+            Expression::float_bin("bin3".to_string()),
+            Expression::float_val(2f64),
+        ),
+        &set_name,
+    );
+    let count = count_results(rs);
+    assert_eq!(count, 1, "FLOAT Test Failed");
+
+    let rs = test_filter(
+        Expression::eq(
+            Expression::blob_bin("bin4".to_string()),
+            Expression::blob_val(format!("{}{}", "blob", 5).into_bytes()),
+        ),
+        &set_name,
+    );
+    let count = count_results(rs);
+    assert_eq!(count, 1, "BLOB Test Failed");
+
+    let rs = test_filter(
+        Expression::eq(
+            Expression::blob_bin("bin4".to_string()),
+            Expression::blob_val(format!("{}{}", "blob", 5).into_bytes()),
+        ),
+        &set_name,
+    );
+    let count = count_results(rs);
+    assert_eq!(count, 1, "BLOB BIN Test Failed");
+
+    let rs = test_filter(
+        Expression::ne(
+            Expression::bin_type("bin".to_string()),
+            Expression::int_val(0),
```

Comment:
Should use `ExpType::NIL` enum.

But also, instead of checking for the bin type `NIL`, you could just use `Expression::bin_exists` here. Maybe better to do a positive check here instead, e.g.

```
    Expression::eq(
        Expression::bin_type("bin".to_string()),
        Expression::int_val(ExpType::STRING)
    )
```

Reply:
> But also, instead of checking for the bin type NIL, you could just use Expression::bin_exists here. Maybe better to do a positive check here instead, e.g.

of course, that would make more sense but this here is to test the bin_type operation. bin_exists should also have one.

Reply:
I understand that. But people also look at the tests to see how to use the client, and this example might be misleading in that regard. That's why I didn't suggest to replace `bin_type` with `bin_exists`, but just to test `bin_type` in a slightly different way.

Reply:
This uses ParticleType and not ExpType btw.